### DISCUSSION TITLE
Please expose keepAliveTimeout and headersTimeout without requiring a custom server

### DISCUSSION THREAD

**Discussion by tabuchid** [vercel/next.js#28642] [Ideas]
### Describe the feature you'd like to request

When deploying behind a load balancer its good to configure the application server with a longer timeout than the load balancer. See [AWS ALB reccomendation](https://docs.aws.amazon.com/elasticloadbalancing/latest/application/application-load-balancers.html#connection-idle-timeout) 

> We also recommend that you configure the idle timeout of your application to be larger than the idle timeout configured for the load balancer. 

AWS defaults a ALB to a 60 second idle timeout so it would be desirable to configure next to have a 61 second timeout.

It looks like this has been discussed in a few places with little to no response from the maintainers:
- https://github.com/vercel/next.js/discussions/16544
- https://github.com/vercel/next.js/discussions/19618
- https://github.com/vercel/next.js/issues/22204

Here is a discussion of a similar issue in node. https://github.com/nodejs/node/issues/27363
and two blogs about people hitting similar issues with ALB's generating 502s using express servers where the solution was to increase the keepAliveTimeout.
- https://adamcrowder.net/posts/node-express-api-and-aws-alb-502/
- https://shuheikagawa.com/blog/2019/04/25/keep-alive-timeout/

### Describe the solution you'd like

Allow setting keepAliveTimeout and headersTimeout.

### Describe alternatives you've considered

Ugly hacks 

> We are doing something similar to patch, just an awk on the bundle after build. As patch replaces the whole file (IIRC) it’s more likely to break things than a regex search and insert. 

https://github.com/vercel/next.js/discussions/19618#discussioncomment-1218511

> We're using patch-package to fix this issue, please make server.keepAliveTimeout a next.js settings 🙏🏻

https://github.com/vercel/next.js/discussions/19618#discussioncomment-1218160

or setting the ALB/ELB to a timeout of 4 seconds
> Heya the only solution I was able to find was to modify the IdleTimeout on the AWS ELB to be 4 seconds, i.e 1 second less than node's default 5 seconds.

https://github.com/vercel/next.js/discussions/16544#discussioncomment-142087

**Comment by tabuchid** [comment]
@timneutkens Since this was converted to a discussion. Do you care to discuss?

**Reply by tabuchid** [reply]
@ijjk @Timer Can we get any insight on this?

**Reply by timneutkens** [reply]
All feature requests are converted to discussions as per the issue template to first gauge if there's any other interest than a single application.

**Reply by rawpixel-vincent** [reply]
Every applications running on an AWS load balancer and that have a fair amount of traffic will get intermittent 502 errors when requesting next.js without changing the keepAliveTimeout.
It's possible a lot of users doesn't detect or doesn't identify the issue, next.js could expose that parameter and point out the AWS alb issue in their documentation.
FWIW it seems to be something that could happen with any load balancer (not specific to aws) that manage a pool of connection for performance.

**Reply by tabuchid** [reply]
> All feature requests are converted to discussions as per the issue template to first gauge if there's any other interest than a single application.

Seems sensible. Is there a threshold in order to be considered? There are a number of issues similar to this one that I've tried to collect into a single ticket here but more may exist. 
Also is this something that could be a community contribution? Where would one look to get started if so?

**Reply by andrewhao** [reply]
We face this issue as well in our infrastructure. Fortunately, most of our apps use custom servers, but I imagine there are a large number of deployments that will go behind ELBs or other load balancers.

**Reply by serle** [reply]
We are also facing this issue, this is a critical fix for running next at any kind of scale

**Reply by DaveyWood** [reply]
We are also facing this issue. We don't use a custom server, and really don't want to add one just to set keep alives and header timeouts.

**Reply by artursOs** [reply]
Same for us here.

**Reply by thomaswienecke** [reply]
Same for us here

**Reply by Especially** [reply]
Experiencing this issue for a massive scale project

**Reply by timneutkens** [reply]
@tabuchid I had another look at this, I'd be happy to review a PR that adds a flag to `next start` to change the timeout as it can't be in `next.config.js` given that it has to be set before next.config.js is loaded. Seems there's only a few 👍 on the feature request though.

**Reply by gsoverini** [reply]
We're facing the same issue here at CleanChoice Energy. We'd like to remove the custom server file, but it's impossible to do until we're able to set `keepAliveTimeout` and `headersTimeout` natively in our NextJs app.

Would it be a bad idea if NextJS read those values from environment variables @timneutkens?

**Reply by timneutkens** [reply]
> Would it be a bad idea if NextJS read those values from environment variables

These changes would only cover `next start` so it would not make sense to have an env var which could be assumed as that it'll apply to every environment.

**Reply by mAAdhaTTah** [reply]
Another +1 on this, we have a custom server we'd like to tear down but this is one of the things blocking us.

**Reply by Miikis** [reply]
@timneutkens I've submitted a PR for this [here](https://github.com/vercel/next.js/pull/35827), whenever you're ready

**Reply by onordgren** [reply]
Any good way of solving this for apps running in standalone mode? Would it be a possibility to be able to set this with a env var that can be picked up in the standalone server.js?

**Reply by Miikis** [reply]
@onordgren I don't think there's any env var for keep-alive timeouts. I have a workaround idea — that I haven't tried yet — but it should work I think:

The custom-server Next.js generates is a bare-bones, Node.js http-server. The idea is to add a post-build step (in your CI, container etc.) that updates node's http-server's keep-alive timeout settings.

So, you have Next.js that generate a standalone server that looks something like:

```js

const NextServer = require('next/dist/server/next-server').default
const http = require('http')
const path = require('path')
process.env.NODE_ENV = 'production'
process.chdir(__dirname)

// Make sure commands gracefully respect termination signals (e.g. from Docker)
// Allow the graceful termination to be manually configurable
if (!process.env.NEXT_MANUAL_SIG_HANDLE) {
  process.on('SIGTERM', () => process.exit(0))
  process.on('SIGINT', () => process.exit(0))
}

let handler

const server = http.createServer(async (req, res) => {
  try {
    await handler(req, res)
  } catch (err) {
    console.error(err);
    res.statusCode = 500
    res.end('internal server error')
  }
})
const currentPort = parseInt(process.env.PORT, 10) || 3000
const hostname = process.env.HOSTNAME || 'localhost'

server.listen(currentPort, (err) => {
  if (err) {
    console.error("Failed to start server", err)
    process.exit(1)
  }
  const nextServer = new NextServer({
    hostname,
    port: currentPort,
    dir: path.join(__dirname),
    dev: false,
    customServer: false,
    conf: {},
  })
  handler = nextServer.getRequestHandler()

  console.log(
    'Listening on port',
    currentPort,
    'url: http://' + hostname + ':' + currentPort
  )
})
```

and you then `sed` (or whatever tool you want to use) something like the below:

```js
const KEEP_ALIVE_TIMEOUT = 70 * 1000;
server.keepAliveTimeout = KEEP_ALIVE_TIMEOUT;
server.headersTimeout = KEEP_ALIVE_TIMEOUT * 1.1;
```

Report back if you try it out :)

**Reply by onordgren** [reply]
Hey @Miikis! Thx for the suggestion :) We are actually already doing something similar in our docker file, like below

`RUN echo ";server.keepAliveTimeout = 620000;" >> ./server.js`

It works fine but it felt a bit hacky and potentially error prone when/if server.js implementation changes.

An env var similar to how `process.env.NEXT_MANUAL_SIG_HANDLE` works would be great for this, at least in my head but maybe there was a reason for not doing this with env vars.

**Reply by Miikis** [reply]
@onordgren Nice! Glad to see it works. I second the env var feature too, tho — I think Next is just waiting for a PR from the community tho at this point lol

**Reply by aNyMoRe0505** [reply]
@Miikis 
Thank you for your PR.
I'd like to confirm whether the PR does not include headersTimeout.
If not, I would like to ask how you addressed the [issue](https://github.com/nodejs/node/issues/27363)
Thank you

**Comment by nelson-lark** [comment]
any momentum on this?  Or any workaround gist?

**Reply by timneutkens** [reply]
As said feel free to open a PR.
https://github.com/vercel/next.js/discussions/28642#discussioncomment-1716954

**Comment by jessethomson** [comment]
For those using standalone mode:

With [this change](https://github.com/vercel/next.js/pull/46052) to NextJS, it looks like you can now use a `KEEP_ALIVE_TIMEOUT` environment variable to set the `keepAliveTimeout` value on the server.

Unfortunately, I don't see this documented anywhere so this was _very_ hard to find.

I should add that from what I can tell, there is still no way to set the `headersTimeout` value (without hacking some code into node_modules or dist). However, it's not clear to me if that is still needed or not. It seems others have mentioned [this commit](https://github.com/nodejs/node/commit/ec1df7b4c9c18673e712203268853c86062ba7cb) on NodeJS and maybe this change in NodeJS means everything is fine _without_ `headersTimeout`? I'm not sure, but would love if someone could shed some light on that.

**Reply by cyounkins** [reply]
I can confirm setting `ENV KEEP_ALIVE_TIMEOUT 61000` in our Dockerfile immediately eliminated our random 502's. 61 sec is appropriate for our environment with a 60 sec timeout in the load balancer.
